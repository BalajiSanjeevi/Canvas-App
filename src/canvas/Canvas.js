import write from '../io/OutputWriter';
import {validateNumbers, validateCommandSetLength} from '../validators/Validator';

const EXPECTED_INPUT_LENGTH = 2;

const BOUNDARY_BORDER_PIXEL_LENGTH = 2;

const LEFT_AND_RIGHT_CANVAS_BORDER_CHAR = "|";

const TOP_AND_BOTTOM_CANVAS_BORDER_CHAR = "-";


/*
 * Canvas class is meant to hold the canvas context which will have the canvas boundary within which we can play around
 * and it has couple of function to help in drawing line and getting the context.
 */
export default class Canvas {

    /*
     * Validates the commandset for the required length and checks whether the command set is all numbers.
     */
    static validate = function (commandSet){
        return  validateCommandSetLength(commandSet, EXPECTED_INPUT_LENGTH) && validateNumbers(commandSet);
    };

    /*
     * Constructor to create the canvas object and initializes the canvas boundary.
     */
    constructor([width, height]){
        this._pixels = [[]];
        this._width = parseInt(width);
        this._height =  parseInt(height);
        this.initializeCanvasBoundary(this._height + BOUNDARY_BORDER_PIXEL_LENGTH, this._width + BOUNDARY_BORDER_PIXEL_LENGTH);
    }

    /*
     * Helps in initializing the canvas boundary.
     */
    initializeCanvasBoundary(height, width){
        this._pixels = [... new Array(height)].map((row, i) => {
            return [... new Array(width)].map((column, j) => {
                if(i === 0 || i === height - 1){
                    return TOP_AND_BOTTOM_CANVAS_BORDER_CHAR;
                } else if(j === 0 || j === width - 1) {
                    return LEFT_AND_RIGHT_CANVAS_BORDER_CHAR;
                }
                return " ";
            });
        });
    }

    /*
     * Exposes the pixels which contains the actual canvas boundary with the state of it.
     */
    get pixels(){
        return this._pixels;
    }

    /*
     *  Calls the write function to print all the rows and columns.
     */
    draw(){
        this._pixels.forEach((row) => {
            write(row.join(""));
        })
    }

    /*
     * Fills the character 'X' from the given startPoint and endPoint index in the given axisY row
     */
    drawHorizontalLine(startPoint, endPoint, axisY){
        // As per Fill function, it fills the character until < endPoint. As we need to include the last position like (<=endPoint). so adding 1 to the endpoint.
        this._pixels[axisY].fill('X', startPoint, endPoint + 1);
    }
    /*
     * Fills the character 'X' on each row from the given startPoint and endPoint index at the given axisX column index.
     */
    drawVerticalLine(startPoint, endPoint, axisX){
        for(var i = startPoint; i <= endPoint; i++){
            this._pixels[i][axisX] = 'X';
        }
    }
}