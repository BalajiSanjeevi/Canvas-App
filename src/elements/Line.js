import Coordinates from '../canvas/Pixel';
import lineValidate from '../validators/LineValidator';

export default class Line {

    constructor(commandSets){
        this._startPoint = new Coordinates(...commandSets);
        this._endPoint = new Coordinates(commandSets[2], commandSets[3]);
    }

    static validate = lineValidate;

    /*
     * Draws either vertical or horizontal line based on the input coordinate pairs.
     */
    draw(canvas){
        if(this.isDrawHorizontalLine()) {
            canvas.drawHorizontalLine(this._startPoint.axisX, this._endPoint.axisX, this._startPoint.axisY);
        } else {
            canvas.drawVerticalLine(this._startPoint.axisY, this._endPoint.axisY, this._startPoint.axisX);
        }
        canvas.draw();
    }

    isDrawHorizontalLine(){
        return this._startPoint.axisY === this._endPoint.axisY;
    }

}