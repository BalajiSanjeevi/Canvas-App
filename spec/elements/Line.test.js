import Line from '../../src/elements/Line';
import Canvas from '../../src/canvas/Canvas';

jest.mock('../../src/io/OutputWriter');

describe("Rectangle", () => {
    let line, canvas;

    beforeEach(() => {
        canvas = new Canvas([10, 4]);
    });

    it("should match snapshot for horizontal line", () => {
        line = new Line([2, 1, 6, 1]);
        line.draw(canvas);
        expect(canvas.pixels).toMatchSnapshot();
    });

    it("should match snapshot for vertical line", () => {
        line = new Line([2, 1, 2, 4]);
        line.draw(canvas);
        expect(canvas.pixels).toMatchSnapshot();
    });

});