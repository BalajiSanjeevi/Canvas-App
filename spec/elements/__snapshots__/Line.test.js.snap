// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rectangle should match snapshot for horizontal line 1`] = `
------------
| XXXXX    |
|          |
|          |
|          |
------------
`;

exports[`Rectangle should match snapshot for vertical line 1`] = `
------------
| X        |
| X        |
| X        |
| X        |
------------
`;
